### DISCUSSION TITLE
Is it possible to url encode with mavo?

### DISCUSSION THREAD

**Discussion by GalinhaLX** [mavoweb/mavo#859] [Q&A] [answered]
Hi

How can I url encode a url with mavo? Does mavo have a function for it?

Let's say I have a property value which is:
`https://mavo.io/docs/storage#url-params`

and I want it to become like this:
`https%3A%2F%2Fmavo.io%2Fdocs%2Fstorage%23url-params`

**Comment by DmitrySharabin** [comment]
For now, there is no way to do it, unfortunately. However, there is an open issue #762 about that. So, feel free to share your thoughts and usecases there.

**Comment by GalinhaLX** [comment, marked as answer]
the same way @LeaVerou is using on her [demo](https://projects.verou.me/homesearch/):

_html_
`<a property="url">[getWebsite(url)]</a>`

_ & javascript_
```
function getWebsite(url) {
	return !!Mavo.value(url) ? new URL(url).host.replace(/^www\./, "") : "(Website)";
}
```

Isn't there a way to use the javascript function called `encodeURIComponent()` ?


Without knowing javascript I tried 
```
function getWebsite(url) {
	return !!Mavo.value(url) ? new URL(url).host.encodeURIComponent() : "(Website)";
}
```

doesn't work, maybe it needs something between ( and )

Or is this a totally wrong way? :(

**Reply by DmitrySharabin** [reply]
Since we can use JS functions in expressions, I would suggest you use the `encodeURIComponent()` function in the needed expression, like so: `[encodeURIComponent(url_to_encode)]`.

In this pen you can see it in action: https://codepen.io/dmitrysharabin/pen/VwQyvBo

Hope it helps! :)

**Reply by LeaVerou** [reply]
Yup, this. There's not much a separate Mavo function would offer here (except handling for lists of values).

**Reply by WebMechanic** [reply]
> Since we can use JS functions in expressions, 

Interesting.
I couldn't find anything about that on https://mavo.io/docs/expressions

**Reply by DmitrySharabin** [reply]
> Interesting. I couldn't find anything about that on https://mavo.io/docs/expressions

Yeah. I believe it's because of it's a bit of an advanced topic (for JS developers). Btw, you might find this interesting too: https://mavo.io/docs/mavoscript.

**Reply by WebMechanic** [reply]
> Btw, you might find this interesting too: https://mavo.io/docs/mavoscript.

thanks, that page doesn't mention it either. It just tells Math functions can be used w/o the `Math.` prefix.
Having the ability to virtually use any arbitrary, custom JS function in expressions should at least be mentioned. 
I had the impression all is "sealed" and I have to go thru the various Mavo API objects to not break some magical render process and data handling.

**Reply by LeaVerou** [reply]
I would swear it was mentioned somewhere in the docs but I cannot find it. @DmitrySharabin do you remember where it is? And if it's not there, we should *definitely* mention it!

**Reply by GalinhaLX** [reply]
Hi 

It's mentioned in [this page](https://mavo.io/get/) in v.0.2.2 , breaking changes:
`Note that if you're using custom JS functions in your expressions, those will continue to work fine, since function calls are perfectly valid MavoScript.`
But I agree it would be better to put it in main documentation

**Reply by DmitrySharabin** [reply]
> I would swear it was mentioned somewhere in the docs but I cannot find it. @DmitrySharabin do you remember where it is? And if it's not there, we should _definitely_ mention it!

You are absolutely right. There was a mention in the docs before we released v0.2.2 (as @GalinhaLX mentioned). I checked the commits to the docs. Those are the ones with the corresponding changes: https://github.com/mavoweb/mavo.io/commit/4fb0087efc02018e675868279b3f632f58eb2745 and https://github.com/mavoweb/mavo.io/commit/803af574b81708f1559f37a00a1a2b922fe85556.

In those commits we removed the following sentences from two sections in the docs:
1. If a Mavo expression is more advanced than MavoScript, it is automatically parsed as JavaScript, allowing experienced web developers to do more with Mavo expressions.
2. Since MavoScript does not support comments, an easy way to force JS mode is a `/* js */` comment in the start or end of the expression.

What do you think if we add something like this?

**If you need something more advanced, you can invoke custom JavaScript functions (both synchronous and asynchronous) in your expressions. Function calls are perfectly valid MavoScript.**

We can add it after the table [here](https://mavo.io/docs/mavoscript#mavoscript-vs-javascript) and [here](https://mavo.io/docs/expressions#mavoscript-vs-javascript). Or anywhere else?

Thoughts?

**Reply by DmitrySharabin** [reply]
> What do you think if we add something like this?
> **If you need something more advanced, you can invoke custom JavaScript functions (both synchronous and asynchronous) in your expressions. Function calls are perfectly valid MavoScript.**
> We can add it after the table [here](https://mavo.io/docs/mavoscript#mavoscript-vs-javascript) and [here](https://mavo.io/docs/expressions#mavoscript-vs-javascript). Or anywhere else?

Ping @LeaVerou. :) What do you think? Can we add this ☝🏻 phrase to the docs?

**Reply by WebMechanic** [reply]
> > We can add it after the table [here](https://mavo.io/docs/mavoscript#mavoscript-vs-javascript) and [here](https://mavo.io/docs/expressions#mavoscript-vs-javascript). 

that's where I looked and would also expect such information :D

**Reply by LeaVerou** [reply]
LGTM. We may also want to use this (URL encoding) as an example.

**Comment by GalinhaLX** [comment]
It SUPER helps ! Thank you so much @DmitrySharabin 

Wow ...even more simple than I thought

I ❤️ Mavo :)

**Comment by karger** [comment]
We may have removed this because there's been some discussion of *not* 
allowing arbitrary js in mavo expressions?

**Reply by DmitrySharabin** [reply]
Yes, IIRC that was the reason.

**Reply by LeaVerou** [reply]
We don't allow arbitrary JS *syntax* in Mavo expressions, but we do allow any arbitrary function.

**Reply by qorncat** [reply]
Ye if you have 🐮 Boy

**Comment by qorncat** [comment]
lets do this
